Explorer/Assets/DCL/Multiplayer/Movement/Tests/TimestampEncoder.cs

```diff
@@ -0,0 +1,95 @@
+        }
+
+        public static (float Timestamp, bool IsMoving) Decode(long encodedValue)
+        {
+            bool isMoving = (encodedValue & (1L << 63)) != 0;
+            long timestamp = (encodedValue >> (63 - BITS)) & 0xFF; // Extract 8-bit timestamp
+
+            // Convert timestamp back to seconds
+            float timestampInSeconds = timestamp * QUANTUM;
+            return (timestampInSeconds, isMoving);
+        }
+    }
+
+    public class Encoder
+{
+    private const int StateBitPosition = 63;
+    private const int TimestampBits = 8;
+    private const int TimestampShift = 55; // Following the state bit
+    private const int XExponentBits = 3;
+    private const int YExponentBits = 5;
+    private const int ZExponentBits = 3;
+    private const int MantissaBits = 5; // Uniform across coordinates
+    private const int XShift = 47; // After timestamp
+    private const int YShift = 39; // After X
+    private const int ZShift = 31; // After Y
+    private const int BitsPerCoordinate = XExponentBits + MantissaBits; // For X and Z
+    private const int YBits = YExponentBits + MantissaBits; // For Y
+
+    public static long Encode(bool isMoving, float timestamp, float x, float y, float z)
+    {
+        // Initial state and timestamp encoding
+        long encoded = (isMoving ? 1L : 0L) << StateBitPosition;
+        encoded |= (long)(timestamp % (1 << TimestampBits)) << TimestampShift;
+
+        // Encode each coordinate
+        encoded |= EncodeCoordinate(x, XExponentBits, MantissaBits) << XShift;
+        encoded |= EncodeCoordinate(y, YExponentBits, MantissaBits) << YShift;
+        encoded |= EncodeCoordinate(z, ZExponentBits, MantissaBits) << ZShift;
+
+        return encoded;
+    }
+
+    private static long EncodeCoordinate(float coordinate, int exponentBits, int mantissaBits)
+    {
+        int bias = (1 << (exponentBits - 1)) - 1;
+        int exponent = Mathf.FloorToInt(Mathf.Log(Mathf.Abs(coordinate), 2)) + bias;
+        float mantissa = (Mathf.Abs(coordinate) / Mathf.Pow(2, exponent - bias)) - 1;
+        long encodedExponent = ((long)exponent & ((1 << exponentBits) - 1)) << mantissaBits;
+        long encodedMantissa = (long)(mantissa * ((1 << mantissaBits) - 1));
+
+        return encodedExponent | encodedMantissa;
+    }
+
+    public static (bool IsMoving, float Timestamp, float X, float Y, float Z) Decode(long encoded)
+    {
+        bool isMoving = (encoded & (1L << StateBitPosition)) != 0;
+        float timestamp = ((encoded >> TimestampShift) & ((1L << TimestampBits) - 1));
+
+        float x = DecodeCoordinate((int)(encoded >> XShift) & ((1 << BitsPerCoordinate) - 1), XExponentBits, MantissaBits);
+        float y = DecodeCoordinate((int)(encoded >> YShift) & ((1 << YBits) - 1), YExponentBits, MantissaBits);
+        float z = DecodeCoordinate((int)(encoded >> ZShift) & ((1 << BitsPerCoordinate) - 1), ZExponentBits, MantissaBits);
+
```

Comment:
Tuples are reference type, please consider structs

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples

Reply:
wait, I was thinking it was Value Tuples. From the link you shared 

> System.ValueTuple types are [value types]. System.Tuple types are [reference types]

and it is ValueTuple indeed
![image](https://github.com/decentraland/unity-explorer/assets/35366872/14271e9e-2df6-48a4-b2ab-2cbb1ea52c23)

Reply:
And here as well - https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-types

> A [value tuple] is a value type, but not a simple type.